dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/AbstractImageRegistry.java

```diff
@@ -240,14 +249,114 @@ public Tool refreshTool(final long toolId, final Long userId, final UserDAO user
         final List<Tool> newDBTools = new ArrayList<>();
         newDBTools.add(toolDAO.findById(tool.getId()));
 
+        List<Tag> toolTags;
         // Get tags and update for each tool
-        List<Tag> toolTags = getTags(tool);
+        if (tool.getRegistry().equals(Registry.DOCKER_HUB.toString())) {
+            toolTags = getTagsDockerHub(tool);
+        } else {
+            toolTags = getTags(tool);
+        }
+
         updateTags(toolTags, tool, sourceCodeRepoInterface, tagDAO, fileDAO, toolDAO, fileFormatDAO);
 
         // Return the updated tool
         return newDBTools.get(0);
     }
 
+    public List<Tag> getTagsDockerHub(Tool tool) {
+        final String repo = tool.getNamespace() + '/' + tool.getName();
+        LOG.info(" ======================= Getting tags for: {}================================", tool.getPath());
+
+        final List<Tag> tags = new ArrayList<>();
+
+        Optional<String> dockerHubResponse = getDockerHubToolAsString(tool);
+
+        Map<String, List<Map<String, List<Map<String, String>>>>> manifestDigestMap = new HashMap<>();
+        Map<String, List<Map<String, String>>> nameMap = new HashMap<>();
+        Map<String, String> error = new HashMap<>();
+
+        if (dockerHubResponse.isPresent()) {
+            Gson gson = new Gson();
+            final String errorJSON = dockerHubResponse.get();
+            final String manifestJSON = dockerHubResponse.get();
+            final String nameJSON = dockerHubResponse.get();
+
+            error = (Map<String, String>)gson.fromJson(errorJSON, error.getClass());
```

Comment:
Alternatively, this would be easier to read if this created typed objects rather than maps of property keys and fields?
Does this match something in https://github.com/spotify/docker-client/tree/14b4464fb7acec25073c8c9c564136188f94562c/src/main/java/com/spotify/docker/client/messages ?